### PR TITLE
[PT-3] - Parallax Periscope Isolation

### PR DESCRIPTION
**Pull request by Dandarprox**
Periscope component isolated and generalization of input image and text for the periscope.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -42,6 +42,7 @@
         "protractor": "~5.4.0",
         "ts-node": "~7.0.0",
         "tslint": "~5.15.0",
-        "typescript": "~3.5.3"
+        "typescript": "~3.5.3",
+        "tslib": "^1.10.0"
```
Comment (roydes) [inline package.json]: Why did you add this dependency? Did you have any problems compiling the app? Any TypeScript Compiler related error? If not,  this is out of the ticket scope.

File: src/app/app.module.ts [hunk 2]
```diff
@@ -3,14 +3,16 @@ import { NgModule } from '@angular/core';
 
 import { AppRoutingModule } from './app-routing.module';
 import { AppComponent } from './app.component';
-import { HomeComponent } from './pages/home-component/home/home.component';
+import { HomeComponent } from './pages/home/home.component';
 import { ScrollAnimationDirective } from './directives/scroll-animation.directive';
+import { PeriscopeComponent } from './components/periscope/periscope.component';
```
Comment (roydes) [inline src/app/app.module.ts]: Please sort the imports in asc mode

File: src/app/components/periscope/periscope.component.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<div class="periscope"
+    [style.backgroundImage]="backgroundImage"
+    [appScrollAnimation]="movePeriscopeBackgroundUp"
+    [checkIfScrolled]="true"
+    [animationExtras]="{ isBeingScrolled: periscopeState, factor: 3 }"
+    (scrolledOver)="setPeriscopeState($event)"
+>
```
Comment (roydes) [inline src/app/components/periscope/periscope.component.html]: this is bad indentation, let the close (>)  in the previous line.

File: src/app/components/periscope/periscope.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  beforeEach(() => {
+    fixture = TestBed.createComponent(PeriscopeComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+});
```
Comment (roydes) [inline src/app/components/periscope/periscope.component.spec.ts]: Improve the coverage up to 95%

File: src/app/components/periscope/periscope.component.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  selector: 'app-periscope',
+  templateUrl: './periscope.component.html',
+  styleUrls: ['./periscope.component.scss']
+})
+export class PeriscopeComponent implements OnInit {
+    @Input() promoTitle: string;
+    @Input() buttonText: string;
+    @Input() imageSrc: string;
+    periscopeState: boolean;
+    backgroundImage: string;
```
Comment (roydes) [inline src/app/components/periscope/periscope.component.ts]: Sort declaration in asc mode

File: src/app/components/periscope/periscope.component.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    constructor() { }
+
+    ngOnInit() {
+        this.backgroundImage = `url(${this.imageSrc})`;
```
Comment (roydes) [inline src/app/components/periscope/periscope.component.ts]: We are storing redundant data here. If the image is just used as a background-image it will be cleaner if we just use one input property.

File: src/app/components/periscope/periscope.component.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    movePeriscopeBackgroundUp (
```
Comment (roydes) [inline src/app/components/periscope/periscope.component.ts]: please change the name of the method to something more general, since it is doing more than moving the background-image up.

File: src/app/pages/home/home.component.html [hunk 8]
```diff
@@ -44,17 +44,11 @@ <h1>Clothing Store</h1>
...
-                [animationExtras]="{ isBeingScrolled: periscopeState, factor: 3 }">
-                <div class="promo-content">
-                    <p class="text">Fall Suits <strong><span>from </span>$99</strong></p>
-                    <a class="shop-now">Shop Now</a>
-                </div>
-            </div>
+            <app-periscope
+                imageSrc="'assets/images/model3.jpg'"
+                promoTitle="Fall suits"
+                buttonText="Shop now"
```
Comment (roydes) [inline src/app/pages/home/home.component.html]: Do not hard code, please move this to the constants files. Eventually, this data will come from services.

**Comment by roydes** [comment]
Please check why is the periscope doing this behavior:
image

File: package.json (hunk 1 above)
Comment (Dandarprox) [reply package.json]: I had some problems using the linter and it was recommended to add that dependency.

File: src/app/app.module.ts (hunk 2 above)
Comment (Dandarprox) [reply src/app/app.module.ts]: Done. But shouldn't be this set in the tslint config file?

Comment (roydes) [reply src/app/app.module.ts]: Yeah, i will add this rule later
